File: task1/src/main/java/framework/elements/BaseElement.java
package framework.elements;

import framework.browser.BrowserFactory;
import framework.utils.MyLogger;
import framework.utils.Reader;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.FluentWait;
import org.openqa.selenium.support.ui.Wait;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.List;

public abstract class BaseElement {
    private String nameElement;
    private By locator;

    public BaseElement(By locator, String nameElement) {
        MyLogger.info("create element " + nameElement);
        this.nameElement = nameElement;
        this.locator = locator;
    }

    public WebElement getElement() {
        MyLogger.info("get element " + nameElement);
        return BrowserFactory.getInstance().findElement(locator);
    }

    public List<WebElement> getElements() {
        MyLogger.info("get elements " + nameElement);
        return BrowserFactory.getInstance().findElements(locator);
    }

    private static WebElement waitElementToBeClickable(WebElement element) {
        WebDriverWait wait = new WebDriverWait(BrowserFactory.getInstance(), Integer.parseInt(Reader.getParametr("timeout")));
        return wait.until(ExpectedConditions.elementToBeClickable(element));
    }

    private static WebElement waitVisibilityElement(WebElement element) {
        Wait<WebDriver> wait = new FluentWait<>(BrowserFactory.getInstance()).withTimeout(Duration.ofSeconds(Integer.parseInt(Reader.getParametr("timeout")))).ignoring(org.openqa.selenium.NoSuchElementException.class);
        return wait.until(ExpectedConditions.visibilityOf(element));
    }

    private static Boolean waitInvsibilityElement(WebElement element) {
        WebDriverWait wait = new WebDriverWait(BrowserFactory.getInstance(), 120);
        return wait.until(ExpectedConditions.invisibilityOf(element));
    }

    public boolean isDisplayedElement() {
        MyLogger.info("element is displayed " + nameElement);

        return waitVisibilityElement(getElement()).isDisplayed();
    }

    public boolean isInvisibleElement() {
        MyLogger.info("element is invisible " + nameElement);
        return waitInvsibilityElement(getElement());
    }

    public boolean isPresent() {
        MyLogger.info("element is present " + nameElement);
        return getElements().size() > 0;
    }

    public void clickElement() {
        MyLogger.info("click on element " + nameElement);
        waitElementToBeClickable(getElement()).click();
    }

    public String getTextFromElement() {
        MyLogger.info("get text from element " + nameElement);
        return getElement().getText();
    }

    public String getAttributeElement(String attribute) {
        MyLogger.info("get attribute " + attribute + " from element");
        return getElement().getAttribute(attribute);
    }
}

File: Aviasales/framework/baseElement.js
const logger = require('./utils/log.util');
require('selenium-webdriver');
const {until} = require('selenium-webdriver');

class BaseElement {

    constructor(browser, name, locator) {
        logger.info(`Create element ${name}`);
        this.browser = browser;
        this.name = name;
        this.locator = locator;
    }

    async findAndClick() {
        logger.info(`Find element and click ${this.name}`);
        let element = await this.findElement().catch((error) => {
            logger.warning(`${error}`);
        });
        await element.click().catch((error) => {
            logger.warning(`${error}`);
        });
    }

    async findElement() {
        logger.info(`Find element ${this.name}`);
        return await this.browser.driver.findElement(this.locator).catch((error) => {
            logger.warning(`${error}`);
        });
    }

    async findElementsFromPage() {
        logger.info(`Find elements ${this.name}`);
        return await this.browser.driver.findElements(this.locator).catch((error) => {
            logger.warning(`${error}`);
        });
    }

    async isDisplayedElement() {
        logger.info(`Is displayed element ${this.name}`);
        return await this.browser.driver.findElement(this.locator).isDisplayed().catch((error) => {
            logger.warning(`${error}`);
        });
    }

    async getText() {
        logger.info(`Get text from element ${this.name}`);
        return await this.browser.driver.findElement(this.locator).getText().catch((error) => {
            logger.warning(`${error}`);
        });
    }

    async getAttributeValue() {
        logger.info(`Set attribute to element ${this.name}`);
        return await this.browser.driver.findElement(this.locator).getAttribute('value').catch((error) => {
            logger.warning(`${error}`);
        });
    }

    async waitOfPresence() {
        logger.info(`Get text from element ${this.name}`);
        return await this.browser.driver.wait(until.elementIsVisible(this.browser.driver.findElement(this.locator))).catch((error) => {
            logger.warning(`${error}`);
        });
    }


     async waitOfHide() {
        logger.info(`Get text from element ${this.name}`);
        await this.browser.driver.wait(until.elementIsNotVisible(this.browser.driver.findElement(this.locator))).catch((error) => {
            logger.warning(`${error}`);
        });
    }
}

module.exports = BaseElement;